lib/routes/account.js

```diff
@@ -333,6 +348,196 @@ module.exports = function (
     },
     {
       method: 'GET',
+      path: '/account/openid/authenticate',
+      handler: function (request, reply) {
+        var id = request.query.identifier
+        if (!isOpenIdProviderAllowed(id)) {
+          log.info({ op: 'Account.openid.authenticate', id: id })
+          return reply.redirect(
+            config.contentServer.url + '/openid?' + qs.stringify(
+              {
+                err: 'This OpenID Provider is not allowed'
+              }
+            )
+          )
+        }
+        openid.authenticate(
+          id,
+          config.openIdVerifyUrl,
+          null, // realm
+          false, // immediate
+          false, // stateless
+          function (err, authUrl) {
+            if (err) {
+              log.error({ op: 'Account.openid.authenticate', err: err })
+              return reply.redirect(
+                config.contentServer.url + '/openid?' + qs.stringify(
+                  {
+                    err: err.message
+                  }
+                )
+              )
+            }
+            reply.redirect(authUrl)
+          },
+          openIdExtensions,
+          false // strict
+        )
+      }
+    },
+    {
+      method: 'GET',
+      path: '/account/openid/login',
+      handler: function (request, reply) {
+        if (!request.url.search) {
+          return reply(
+          '<?xml version="1.0" encoding="UTF-8"?>\n'
+          + '<xrds:XRDS xmlns:xrds="xri://$xrds" xmlns="xri://$xrd*($v*2.0)"><XRD>'
+          + '<Service xmlns="xri://$xrd*($v*2.0)">'
+          + '<Type>http://specs.openid.net/auth/2.0/return_to</Type>'
+          + '<URI>' + config.openIdVerifyUrl + '</URI>'
+          + '</Service></XRD></xrds:XRDS>'
+          ).type('application/xrds+xml')
+        }
+        log.info({ op: 'Account.openid', url: request.url })
+        openid.verifyAssertion(
+          url.format(request.url),
+          function (err, assertion) {
+            if (err || !assertion || !assertion.authenticated) {
+              log.warn({ op: 'Account.openid', err: err, assertion: assertion })
+              return reply.redirect(
+                config.contentServer.url + '/openid?err=Unknown%20Account'
+              )
+            }
+            var id = assertion.claimedIdentifier
+            var locale = request.app.acceptLanguage
+
+            if (!isOpenIdProviderAllowed(id)) {
```

Comment:
Is there any way to perform this check before `verifyAssertion` above?  Otherwise ISTM we could be caused to make an outbound request to any arbitrary domain.  Which may not be a problem, but if we don't have to then it's probably worth not doing it, from a security POV.